src/middlewares/csrf.ts

```diff
@@ -0,0 +1,121 @@
+import type { FreshContext } from "../context.ts";
+import { HttpError } from "../error.ts";
+
+/**
+ * Determines whether a given origin is allowed for CSRF protection.
+ *
+ * @param origin - The origin of the incoming request.
+ * @param context - The FreshContext object for the current request.
+ * @returns A boolean indicating if the origin is allowed.
+ */
+export type IsAllowedOriginHandler = (
+  origin: string,
+  context: FreshContext,
+) => boolean;
+
+/**
+ * Options for configuring the CSRF protection middleware.
+ *
+ * @property origin - Specifies the allowed origin(s) for requests. Can be a string, an array of strings, or a custom handler function.
+ *   - string: A single allowed origin.
+ *   - string[]: Multiple allowed origins.
+ *   - IsAllowedOriginHandler: A function to determine if an origin is allowed.
+ */
+export interface CsrfOptions {
+  origin?: string | string[] | IsAllowedOriginHandler;
+}
+
+function isSafeMethod(method: string): boolean {
+  return method === "GET" || method === "HEAD";
+}
+
+function isRequestedByFormElement(contentType: string): boolean {
+  const contentTypeLower = contentType.toLowerCase();
+
+  return contentTypeLower === "application/x-www-form-urlencoded" ||
+    contentTypeLower === "multipart/form-data" ||
+    contentTypeLower === "text/plain";
+}
+
+/**
+ * CSRF Protection Middleware for Fresh.
+ * Adapted from Hono's CORS middleware.
+ * Would like to express my gratitude to all the developers of Hono.
+ *
+ * @param options Options for the CSRF protection middleware.
+ * @returns The middleware handler function.
+ *
+ * @example Specifying origins with using `origin` option: default
+ * ```ts
+ * const app = new App<State>()
+ *
+ * app.use(csrf())
+ * ```
+ *
+ * @example Specifying origins with using `origin` option: string
+ * ```ts
+ * app.use(csrf({ origin: 'https://myapp.example.com' }))
+ *
+ * // string[]
+ * app.use(
+ *   csrf({
+ *     origin: ['https://myapp.example.com', 'http://development.myapp.example.com'],
+ *   })
+ * )
+ * ```
+ *
+ * @example Specifying origins with using `origin` option: function
+ * It is strongly recommended that the protocol be verified to ensure a match to `$`.
+ * You should *never* do a forward match.
```

Comment:
```suggestion
 * @example Specifying more complex origins
```
I think any explanation should be clear, concise and lie between the example title and the snippet.